src/java.base/share/classes/java/util/ReversibleCollection.java

```diff
@@ -0,0 +1,142 @@
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package java.util;
+
+/**
+ * A linear collection that has a well-defined order and that is reversible.
+ * Provides methods for a reverse-ordered view as well as convenient access
+ * to elements at both ends.
+ *
+ * @param <E> the type of elements in this collection
+ * @since XXX
+ */
+public interface ReversibleCollection<E> extends Collection<E> {
+    /**
+     * Returns a reversed-order view of this collection. If the implementation
+     * permits modifications to this view, the modifications "write through"
+     * to the underlying collection. Depending upon the implementation's
+     * concurrent modification policy, changes to the underlying collection
+     * may be visible in this reversed view.
+     * @return a reversed-order view of this collection
+     */
+    ReversibleCollection<E> reversed();
+
+    /**
+     * Adds an element at the front of this collection (optional operation).
+     * @param e the element to be added
+     * @throws NullPointerException if the specified element is null and this
+     *         collection does not permit null elements
+     * @throws UnsupportedOperationException if this collection implementation
+     *         does not support this operation
+     * @implSpec
+     * The default implementation in this class throws UnsupportedOperationException.
+     */
+    default void addFirst(E e) {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Adds an element at the end of this collection (optional operation).
+     * Synonymous with add(E).
+     * @param e the element to be added.
+     * @throws NullPointerException if the specified element is null and this
+     *         collection does not permit null elements
+     * @throws UnsupportedOperationException if this collection implementation
+     *         does not support this operation
+     * @implSpec
+     * The default implementation in this class throws UnsupportedOperationException.
+     */
+    default void addLast(E e) {
+        throw new UnsupportedOperationException();
+    }
+
```

Comment:
May we have the `addAll` combined with `addFirst`/`addLast`? For instance, if we can add a collection at the first of an array list, it would be somehow more efficient than doing so for each element (only one system arraycopy operation)

Reply:
Interesting. I guess, `addAllFirst` and `addAllLast`. I don't think I want to include these as part of ReversibleCollection, since they really are new APIs. They certainly apply to List, but they don't apply to SortedSet (which doesn't support add at all), they don't seem to provide any advantage to LinkedHashSet, and there doesn't seem to be a way to optimize `addAllFirst` for a Deque in general. (ArrayDeque could clearly be optimized for bulk-add-at-front, but there's nothing in the Deque interface that supports that.) Note that `addAllLast` is mostly the same as `addAll` for List and Deque.

It might be useful as an independent proposal, though I'm not sure how frequently this comes up.

If what you're adding is reversible, you can do this:

    src.reversed().forEach(dest::addFirst);

Of course this is horribly inefficient if `dest` is an ArrayList, but in that case you can do this:

    dest.addAll(0, src);

Reply:
It sounds good enough as long as we have a proper and optimized `addAll` for the reversed collection. Thanks!